Auto-generate mixin configs

Phoenix-Starlight opened the pull request:
Resolves #297 
1 (or 2?) ugly hack since AP output API doesn't have anything for resources
On the positive side, the directory scanning hack is dead.
- [x] Auto-generate mixin configs
- [x] Integrate AP
- [x] ~~Get victims to test whether or not it explodes~~
- [x] New: Emit compile-time errors for mixins that target client side classes but are not annotated with `@ClientOnlyMixin`
- [x] Fix remaining compile-time errors

Phoenix-Starlight committed:
Create mixin config AP
- Takes two compiler arguments: -ArootProject.name and an optional -Aproject.name

Phoenix-Starlight committed:
Integrate AP

Phoenix-Starlight committed:
Spotless

embeddedt commented:
I can test it easily by checking that it generates the same JSON as the previous approach. No victims required.

Phoenix-Starlight commented:
RemapJar task originally didn't take the JSON. (Hence the hack into resource sourceset)

Phoenix-Starlight committed:
Cleanup

Phoenix-Starlight committed:
Split annotations to separate project
Use fabric mixin dependency
Enable preview features for annotation processor only

Phoenix-Starlight committed:
Emit compile-time errors for mislabeled mixins that targets client side classes

Phoenix-Starlight committed:
Fix NPEs
Apparently class retained annotations are not available in AnnotatedConstruct#getAnnotation

Phoenix-Starlight committed:
Cleanup

Phoenix-Starlight committed:
Fix compile-time errors

Phoenix-Starlight committed:
Fix file naming

Phoenix-Starlight committed:
Exclude tests

Phoenix-Starlight committed:
Add platform support and rework detection logic

Phoenix-Starlight committed:
Fix platform support
- Prefer TypeElements over TypeMirrors
- Add debug option to log classes that couldn't be loaded
- Reorganize exception logging

Phoenix-Starlight committed:
Fix platform detection again
Halt AP on exception

Phoenix-Starlight committed:
Enable debug

Phoenix-Starlight committed:
Revert debug

Phoenix-Starlight committed:
Brute force all markers

Phoenix-Starlight committed:
Fix overloads

Phoenix-Starlight committed:
Debug again

Phoenix-Starlight committed:
Bring in platform classes

Phoenix-Starlight committed:
Retry

Phoenix-Starlight committed:
Add in neoforge maven

Phoenix-Starlight committed:
Rework platform detection logic again

Phoenix-Starlight committed:
Don't rely on enums being present

Phoenix-Starlight committed:
Runtime only classes

Phoenix-Starlight committed:
Disable debug

This reverts commit 1105830e2ee64b5b33bd481e50fd1fd7628e2e86.

Phoenix-Starlight committed:
Prevent CME

Phoenix-Starlight committed:
Fix Forge mixins

Phoenix-Starlight committed:
Remove brute force debug

Phoenix-Starlight committed:
Shadow relocate Fabric Mixin and intentionally override Mixin AP

Phoenix-Starlight committed:
Don't assume common mixins is always present

Phoenix-Starlight committed:
Rework shadowing

Phoenix-Starlight committed:
Unify shadow version
Dependency rework
Remove class shunting hack

Phoenix-Starlight committed:
Fix property substitution

Phoenix-Starlight committed:
Unshadow mixin processor
There is no way to prevent the relocator transformation so we depend on the unshadowed project that pulls the shadowed project
Cleanup client validator shadow jar

Phoenix-Starlight committed:
Depend on shadow config
Fix shadowed unbundled dependencies

Phoenix-Starlight committed:
Manually shadow Mixin

Phoenix-Starlight committed:
Unshadow mixin again, wean off dependency

Phoenix-Starlight committed:
Reunify AP

Phoenix-Starlight committed:
Customize reshadowing of dependencies
Refactor AP detection

Phoenix-Starlight committed:
Tidy up

Phoenix-Starlight commented:
Ready. (Merge conflicts incoming for later branches...)

Phoenix-Starlight committed:
In the name of type-safety
More cleanup and change equality checks

embeddedt commented on forge/src/main/java/org/embeddedt/modernfix/forge/mixin/perf/async_locator/EyeOfEnderAccess.java:
```diff
@@ -1,11 +1,13 @@
 import org.spongepowered.asm.mixin.gen.Accessor;
 
 @Mixin(EyeOfEnder.class)
+@ClientOnlyMixin
```
The async locator mixins should not be client-only.

Phoenix-Starlight committed:
Add annotation to ignore mixins

Phoenix-Starlight committed:
Remove throws declaration

Phoenix-Starlight committed:
Spotless

embeddedt commented on buildSrc/src/main/groovy/modernfix.mod-common-conventions.gradle:
```diff
@@ -16,37 +16,26 @@ dependencies {
-            expand replacements
+tasks.withType(JavaCompile) {
+    options.fork = true
+    options.forkOptions.jvmArgs << '--enable-preview'
```
Why does the mod itself need to be compiled with `--enable-preview`? It conflicts with the commit description, which says it's only needed for the AP: https://github.com/embeddedt/ModernFix/commit/4645c2622742d62ecf9fd08028897edb40512be3

Phoenix-Starlight replied:
So the AP itself is compiled with preview... but you need to enable preview features on the JVM to run the AP. This hunk specifically enables it for the compilation process/AP but not the code being compiled.
See https://github.com/gradle/gradle/issues/12680

embeddedt reviewed (changes requested):
Very good work; I'm very happy with this. Tested locally and the generated mixin list is identical to my original hack, except it properly splits client mixins into the client list.

I just have one last question before merging.

(At some point, we might consider moving this into a standalone plugin so that it can be used by other projects.)
